### PR TITLE
specialized constructors for singly-exposed DOM types

### LINKED ISSUE #14071: Specialize constructors for DOM types that are only exposed to a single global
Currently, a [WebIDL file](https://github.com/servo/servo/blob/master/components/script/dom/webidls/Range.webidl) with a `[Constructor]` annotation generates Rust code like this:
```rust
unsafe extern fn _constructor(cx: *mut JSContext, argc: u32, vp: *mut JSVal) -> bool {
    return wrap_panic(|| {
        let global = GlobalScope::from_object(JS_CALLEE(cx, vp).to_object());
        let args = CallArgs::from_vp(vp, argc);
        let result: Result<Root<Range>, Error> = Range::Constructor(&global);
```
(where `Range` is the name of the WebIDL interface in question). This means that the `Constructor` method needs to take a `&GlobalScope` argument, so that it can be constructed inside of any JS global environment (eg. a window, a worker, a service worker, etc.).

For interfaces are only exposed to a single global (ie. they have no `[Exposed]` annotation that [explicitly exposes them to multiple globals](https://github.com/servo/servo/blob/7c6793fc28423e518a0ceaae1176048fcb2caa3a/components/script/dom/webidls/TextEncoder.webidl#L6)), we can generate more specialized code instead. We would like to generate:
```rust
unsafe extern fn _constructor(cx: *mut JSContext, argc: u32, vp: *mut JSVal) -> bool {
    return wrap_panic(|| {
        let global = GlobalScope::from_object(JS_CALLEE(cx, vp).to_object());
        let global = global.downcast::<Window>().unwrap();
        let args = CallArgs::from_vp(vp, argc);
        let result: Result<Root<Range>, Error> = Range::Constructor(&global);
```
This will require the `Constructor` method to accept a `&Window` argument instead (and will therefore require updating parts of the code that benefit from this).

The Python code that generates this Rust code lives in `CGClassConstructorHook` in `components/script/dom/bindings/codegen/CodegenRust.py`. We can use `descriptor.interface.exposureSet` to get the list of global names that an interface is exposed in, so we can check if the list only has one entry and downcast to the name present in that list.

There should be no need to run automated tests for these changes; if the resulting code changes build, that's good enough for a PR!

Note: the generated code can be found in `target/debug/build/script-[hash]/out/Bindings/[Interface]Binding.rs`.

### PR DESCRIPTION
**Pull request by ajnirp**
---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #14071 (github issue number if applicable).

- [ ] There are tests for these changes OR
- [X] These changes do not require tests because these are codegen changes and the generated code builds successfully

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14160)

### CODE REVIEW

File: components/script/dom/bindings/codegen/CodegenRust.py [hunk 1]
```diff
@@ -5247,12 +5247,18 @@ def __init__(self, descriptor, constructor=None):
...
-        preamble = CGGeneric("""\
-let global = GlobalScope::from_object(JS_CALLEE(cx, vp).to_object());
-let args = CallArgs::from_vp(vp, argc);
-""")
+        preamble = """let global = GlobalScope::from_object(JS_CALLEE(cx, vp).to_object());\n"""
+        if len(self.exposureSet) == 1:
+            globalName = list(self.exposureSet)[0]
+            downcast = """let global = global.downcast::<dom::{}::{}>().unwrap();\n"""
+            downcast = downcast.format(globalName.lower(),
+                                       globalName[0].upper() + globalName[1:])
```
Comment (nox) [inline components/script/dom/bindings/codegen/CodegenRust.py]: Could you explain what these two lines are supposed to do? Also you may want to look at `MakeNativeName` in Configuration.py.

File: components/script/dom/bindings/codegen/CodegenRust.py [hunk 2]
```diff
@@ -3170,7 +3170,7 @@ def __init__(self, errorResult, arguments, argsPre, returnType,
 
         if isFallible:
             if static:
-                glob = "&global"
+                glob = "global.upcast::<GlobalScope>()"
```
Comment (nox) [inline components/script/dom/bindings/codegen/CodegenRust.py]: Why is this needed?

File: components/script/dom/webglcontextevent.rs [hunk 3]
```diff
@@ -82,6 +83,8 @@ impl WebGLContextEvent {
 
         let cancelable = EventCancelable::from(init.parent.cancelable);
 
+        let global = window.upcast::<GlobalScope>();
```
Comment (nox) [inline components/script/dom/webglcontextevent.rs]: `WebGLContextEvent::new` should probably be made to take a `&Window` too.

File: components/script/dom/extendableevent.rs [hunk 4]
```diff
@@ -41,9 +42,10 @@ impl ExtendableEvent {
         ev
     }
 
-    pub fn Constructor(global: &GlobalScope,
+    pub fn Constructor(worker: &ServiceWorkerGlobalScope,
                        type_: DOMString,
                        init: &ExtendableEventBinding::ExtendableEventInit) -> Fallible<Root<ExtendableEvent>> {
+        let global = worker.upcast::<GlobalScope>();
```
Comment (nox) [inline components/script/dom/extendableevent.rs]: `ExtendableEvent::new` should take a `&Window`.

File: components/script/dom/extendablemessageevent.rs [hunk 5]
```diff
@@ -46,10 +47,11 @@ impl ExtendableMessageEvent {
         ev
     }
 
-    pub fn Constructor(global: &GlobalScope,
+    pub fn Constructor(worker: &ServiceWorkerGlobalScope,
                        type_: DOMString,
                        init: &ExtendableMessageEventBinding::ExtendableMessageEventInit)
                        -> Fallible<Root<ExtendableMessageEvent>> {
+        let global = worker.upcast::<GlobalScope>();
```
Comment (nox) [inline components/script/dom/extendablemessageevent.rs]: `ExtendableMessageEvent::new` should take a `&Window`.

File: components/script/dom/dommatrixreadonly.rs [hunk 6]
```diff
@@ -40,12 +42,14 @@ impl DOMMatrixReadOnly {
     }
 
     // https://drafts.fxtf.org/geometry-1/#dom-dommatrixreadonly-dommatrixreadonly
-    pub fn Constructor(global: &GlobalScope) -> Fallible<Root<Self>> {
+    pub fn Constructor(window: &Window) -> Fallible<Root<Self>> {
```
Comment (nox) [inline components/script/dom/dommatrixreadonly.rs]: `DOMMatrix` and friends are supposed to be `Exposed=(Window,Worker)`.

**Review by nox — CHANGES_REQUESTED** [review]
Thanks for your contribution!

File: components/script/dom/bindings/codegen/CodegenRust.py (hunk 1 above)
Comment (ajnirp) [inline components/script/dom/bindings/codegen/CodegenRust.py]: This is for the case when a DOM type is exposed to exactly one global. In that case, we downcast `global` from `GlobalScope` to that particular struct.

Thank you for pointing out `MakeNativeName`! I have fixed it.

File: components/script/dom/webglcontextevent.rs (hunk 3 above)
Comment (ajnirp) [inline components/script/dom/webglcontextevent.rs]: Done!

File: components/script/dom/extendableevent.rs (hunk 4 above)
Comment (ajnirp) [inline components/script/dom/extendableevent.rs]: I think this should be `&ServiceWorkerGlobalScope`, based on `ExtendableEvent.webidl` Updated!

File: components/script/dom/extendablemessageevent.rs (hunk 5 above)
Comment (ajnirp) [inline components/script/dom/extendablemessageevent.rs]: Likewise, I think this should be `&ServiceWorkerGlobalScope` as well. Updated!

File: components/script/dom/bindings/codegen/CodegenRust.py [hunk 7]
```diff
@@ -5247,12 +5247,18 @@ def __init__(self, descriptor, constructor=None):
...
 
     def definition_body(self):
-        preamble = CGGeneric("""\
-let global = GlobalScope::from_object(JS_CALLEE(cx, vp).to_object());
-let args = CallArgs::from_vp(vp, argc);
-""")
+        preamble = """let global = GlobalScope::from_object(JS_CALLEE(cx, vp).to_object());\n"""
+        if len(self.exposureSet) == 1:
+            globalName = list(self.exposureSet)[0]
+            downcast = """let global = global.downcast::<dom::{}::{}>().unwrap();\n"""
```
Comment (nox) [inline components/script/dom/bindings/codegen/CodegenRust.py]: I think you can just do:

```python
let downcast = "global.downcast::<%s>().unwrap();" % descriptor.path
```

File: components/script/dom/extendablemessageevent.rs (hunk 5 above)
Comment (nox) [inline components/script/dom/extendablemessageevent.rs]: Oh right! :)

File: components/script/dom/webidls/DOMMatrix.webidl [hunk 8]
```diff
@@ -12,8 +12,8 @@
 
 [Constructor,
 // Constructor(DOMString transformList),
-Constructor(sequence<unrestricted double> numberSequence)
-// Exposed=(Window,Worker)
+Constructor(sequence<unrestricted double> numberSequence),
+Exposed=(Window,Worker)
```
Comment (nox) [inline components/script/dom/webidls/DOMMatrix.webidl]: Could you do that in a first separate commit?

**Review by nox — CHANGES_REQUESTED** [review]

**Comment by nox** [comment]
Carried in #14376, thanks for your contribution!

**Pull request closed without merging**